plugins/base/plugin_test.go

```diff
@@ -0,0 +1,176 @@
+package shared
+
+import (
+	"testing"
+
+	pb "github.com/golang/protobuf/proto"
+	plugin "github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/hashicorp/nomad/plugins/base/proto"
+	"github.com/hashicorp/nomad/plugins/shared/hclspec"
+	"github.com/stretchr/testify/require"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/msgpack"
+	"google.golang.org/grpc"
+)
+
+var (
+	// testSpec is an hcl Spec for testing
+	testSpec = &hclspec.Spec{
+		Block: &hclspec.Spec_Object{
+			&hclspec.Object{
+				Attributes: map[string]*hclspec.Spec{
+					"foo": &hclspec.Spec{
+						Block: &hclspec.Spec_Attr{
+							&hclspec.Attr{
+								Type:     "string",
+								Required: false,
+							},
+						},
+					},
+					"bar": &hclspec.Spec{
+						Block: &hclspec.Spec_Attr{
+							&hclspec.Attr{
+								Type:     "number",
+								Required: true,
+							},
+						},
+					},
+					"baz": &hclspec.Spec{
+						Block: &hclspec.Spec_Attr{
+							&hclspec.Attr{
+								Type: "bool",
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+)
+
+// testConfig is used to decode a config from the testSpec
+type testConfig struct {
+	Foo string `cty:"foo" codec:"foo"`
+	Bar int64  `cty:"bar" codec:"bar"`
+	Baz bool   `cty:"baz" codec:"baz"`
+}
+
+func TestBasePlugin_PluginInfo_GRPC(t *testing.T) {
+	t.Parallel()
+	require := require.New(t)
+
+	const (
+		apiVersion    = "v0.1.0"
+		pluginVersion = "v0.2.1"
+		pluginName    = "mock"
+	)
+
+	knownType := func() (*PluginInfoResponse, error) {
+		info := &PluginInfoResponse{
+			Type:             PluginTypeBase,
+			PluginApiVersion: apiVersion,
+			PluginVersion:    pluginVersion,
+			Name:             pluginName,
+		}
+		return info, nil
+	}
+	unknownType := func() (*PluginInfoResponse, error) {
+		info := &PluginInfoResponse{
+			Type:             "bad",
+			PluginApiVersion: apiVersion,
+			PluginVersion:    pluginVersion,
+			Name:             pluginName,
+		}
+		return info, nil
+	}
+
+	mock := &MockPlugin{
+		PluginInfoF: knownType,
+	}
+
+	conn, server := plugin.TestGRPCConn(t, func(s *grpc.Server) {
```

Comment:
You might want to use [TestPluginGRPCConn](https://godoc.org/github.com/hashicorp/go-plugin#TestPluginGRPCConn) and dispense a basePluginClient.